Relay support (fixes #1364)

AudriusButkevicius opened the pull request:
The first changeset does:

1. Change listen addresses to URIs everywhere
2. Break out connectionSvc to support listeners and dialers based on schema (not necessary, but nice to have)
3. Add relay announcement and lookups to discovery service

Things left to do:
1. Add a subservice to connectionSvc for dealing with relay addresses
2. Keep dialing non-relay addresses, and disconnect existing relay connections if we've managed to get a direct connection.

calmh commented:
Looks reasonable at first glance. Also very incompatible with v0.11. :) Should we keep a `relay` branch to merge relay pr:s to, so we can keep small stuff on the master branch and drop another v0.11 from that one? That way we can also avoid having to review everything in a big blob :) Then the relay branch becomes v0.12.0 when it's ready?

AudriusButkevicius commented:
I guess this is a PR for the sake of review, rather than being merged.
I am fine to have a v0.12 branch and merge that there, or vica versa, have a v0.11 branch and merge this into master once its done.

The first commit will probably stay as a big blob as it will be a hassle to split it up now.
The rest will be smaller ones which will be easier to follow.

AudriusButkevicius commented:
Also, you could make this part of v0.11 if you wedged in the old packet support to disosrv

calmh commented on internal/discover/discover.go:
```diff
@@ -184,75 +192,108 @@ func (d *Discoverer) ExtAnnounceOK() map[string]bool {
 
-	if len(cached) > 0 {
+	relays := make([]string, 0, len(cachedRelays))
+	for i := range cachedRelays {
```
`index out of bounds` (want `make([]string, len(cachedRelays))` above)

calmh commented on internal/discover/discover.go:
```diff
@@ -184,75 +192,108 @@ func (d *Discoverer) ExtAnnounceOK() map[string]bool {
+
+		relays = addressesSortedByLatency(availableRelays)
+		cachedRelays := []CacheEntry{}
+		for _, relay := range relays {
```
Could be `cachedRelays := make([]CacheEntru, len(relays)); for i := relay {` as above

calmh commented on internal/discover/discover.go:
```diff
@@ -363,19 +412,18 @@ func (d *Discoverer) registerDevice(addr net.Addr, device Device) bool {
-		var deviceAddr string
-		if len(a.IP) > 0 {
-			deviceAddr = net.JoinHostPort(net.IP(a.IP).String(), strconv.Itoa(int(a.Port)))
-		} else if addr != nil {
```
Does this remove the handling for taking the packet source address when the address in the packet is empty? (Used by local discovery)

AudriusButkevicius replied:
Good spot.
Fixed.

calmh commented on internal/discover/discover.go:
```diff
@@ -424,30 +472,103 @@ func (d *Discoverer) filterCached(c []CacheEntry) []CacheEntry {
+			}
+			relays[i].Latency = int32(latency / time.Millisecond)
+		} else {
+			l.Debugf("Failed to get relay %s latency %s", addr, err)
```
Does this mean unreachable relays get zero latency and are preferred?

AudriusButkevicius replied:
The default value is `Latency: int32(time.Hour / time.Millisecond)`

calmh commented on internal/osutil/ping.go:
```diff
@@ -0,0 +1,53 @@
+// Ping returns the latency to the given host
+// Raw sockets require root/administrator privileges, hence we work around
+// that by using the OS binary.
+func Ping(host string) (time.Duration, error) {
```
Could we just time a tcp connect with the lobby port instead?

AudriusButkevicius replied:
And divide that by 3 or something?

AudriusButkevicius replied:
Now replaced with TcpPing, which sadly leaves state around, but I guess that's the best we can do without dropping down to raw sockets.

calmh commented on cmd/syncthing/connections.go:
```diff
@@ -374,3 +325,19 @@ func (s *connectionSvc) CommitConfiguration(from, to config.Configuration) bool
+	if err = conn.SetNoDelay(false); err != nil {
+		l.Infoln(err)
+	}
+	if err = conn.SetKeepAlivePeriod(60 * time.Second); err != nil {
```
(re: IRC discussion yesterday) Huh, apparently we do at least try to set these and I don't recall any bugs in the Infoln()s that should have been printed if these failed... Would be interesting to some day look at a tcpdump of an idle connections and see if this actually does what it says on the tin on all relevant platforms... Then we could skip the ping/pong stuff potentially.

calmh commented on cmd/syncthing/connections_tcp.go:
```diff
@@ -0,0 +1,95 @@
+		return nil, err
+	}
+
+	conn, err := net.DialTCP("tcp", nil, raddr)
```
maybe ResolveTCPAddr + DialTCP could just be a regular `Dial("tcp", ...)` followed by a type assertion `tcpConn := conn.(*net.TCPConn)` on the result... Not sure if that's better or worse though. It's fewer lines :)

calmh commented on cmd/syncthing/connections_tcp.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func tcpListener(uri *url.URL, tlsCfg *tls.Config, conns chan<- *tls.Conn) {
+	tcaddr, err := net.ResolveTCPAddr("tcp", uri.Host)
```
As above, if it makes sense and makes things clearer...

calmh commented on internal/config/testdata/v11.xml:
```diff
@@ -0,0 +1,12 @@
+<configuration version="10">
```
Not actually a v11 config, not that there are any changes in here anyway... ;) Maybe we should include some options in our example configs to make sure they parse correctly.

AudriusButkevicius replied:
That's proven by defaults and override defaults.

AudriusButkevicius commented:
This now actually works, if you use the new discovery server together with the new relaysrv.

AudriusButkevicius commented:
I guess lets get it reviewed before I put in the effort of fixing tests etc.
Feels like the connection type should be part of the model (so that it could be reported in the UI?)

calmh commented:
Looked through it quickly, only on the phone right now... Looks mostly fine. Re your point above, we could have AddConnection take an interface that matches protocol.Connection plus a ConnectionType(), and wrap it to add info about the connection type - `type typedConn struct { protocol.Connection; ctype connectionType }` or something.

calmh commented on cmd/syncthing/connections.go:
```diff
@@ -29,7 +39,44 @@ type connectionSvc struct {
+}
+
+const (
+	connectionTypeBasicAccept connectionType = 1 << iota
```
bit shiftery implies something could be `connectionTypeBasicDial | connectionTypeRelayAccept` or something which does not seem to be the case?

AudriusButkevicius replied:
There is a .IsDirect for conntype, or should be, on the phone atm.

calmh commented on cmd/syncthing/relays.go:
```diff
@@ -0,0 +1,167 @@
+	if debugNet {
+		l.Debugln("Relay service starting")
+	}
+	s.Supervisor.ServeBackground()
```
I'm not sure I like this. Would it make more sense to just `Add()` the relevant stuff to itself in the constructor and let whoever created the relayService start it?

AudriusButkevicius replied:
Yeah this whole starting and stopping thing is hard. 
Checking for nil channels etc just seems like a hack. 

Though there is a teardown phase, so I guess I cannot just wrap it in a serviceFunc, plus it implements the config comitter so can't be a stateless service func.

calmh commented on cmd/syncthing/relays.go:
```diff
@@ -0,0 +1,167 @@
+		uri, err := url.Parse(addr)
+		if err != nil {
+			if debugNet {
+				l.Debugln("Failed to parse relay address", addr, err)
```
This would probably fit better in `VerifyConfiguration` as we can actually tell they are unparseable there

AudriusButkevicius replied:
Agree.

calmh commented on cmd/syncthing/relays.go:
```diff
@@ -0,0 +1,167 @@
+		}
+	}
+
+	for uri, token := range s.tokens {
```
Could this be better in `Stop()` or as something deferred on `Serve()` or so?

AudriusButkevicius replied:
This is to deal with changes while the service is running.

calmh replied:
Yeah, I was thinking of the pattern where Serve() can do restarts by itself, if it crashes or is signalled by a socket close or channel close. So we could do the relevant reconfig, then signal stop and Serve() does it's thing in a possibly deferred cleanup function then starts up again with the new config.

AudriusButkevicius replied:
I would find that more confusing, as it's the supervisors job to deal with that, isn't it?

calmh replied:
No, I mean like in gui.go for example. Do the prep work in CommitConfiguration, then tell Serve to exit. Serve does whatever cleanup necessary on the way out - as it would in a panic or clean shutdown. The supervisor restarts it. Everyone is happy.

AudriusButkevicius replied:
I guess that means storing a slice of addresses, and locking?

calmh replied:
I don't know, I didn't think that closely, sorry. :) I just felt the teardown would feel better in or after serve than in commit config

calmh commented on internal/osutil/ping.go:
```diff
@@ -0,0 +1,27 @@
+	if conn != nil {
+		conn.Close()
+	}
+	return time.Now().Sub(start), err
```
`time.Since(start)` :)

AudriusButkevicius commented:
Have you had a peek at the packages this code uses? Ala srv client and protocol?

calmh commented:
Ah, no, sorry. I'm back in normal connectivity now though, will do so, hopefully this evening. :)

AudriusButkevicius commented:
I believe I've addressed everything in this PR.

Reviews and potentially changes still required in discosrv and relaysrv repos.

AudriusButkevicius commented:
Also, given we're changing discovery causing incompatibilities, my hands are itching to roll out TLS based one to prevent DoS. (or atleast a new type of discovery client which supports relays and uses TLS)

AudriusButkevicius commented:
Rest assured that this hasn't been excessively tested though ;)
It connects as far as I know.

calmh commented:
I did look through the relaysrv packages and they looked fine to me - some stylistic nits I could pick, but nothing structural or *wrong* to complain about. The above is probably fine but let me take one more pass through just for the sake of it - long flights tomorrow, good for reviewing code. :) Discosrv I haven't checked yet, will do.

Zillode commented:
@st-jenkins run integration tests for windows please

Zillode commented:
@st-jenkins run integration tests for windows please

AudriusButkevicius commented:
@st-jenkins run integration tests please

AudriusButkevicius commented:
@st-jenkins run integration tests please

AudriusButkevicius commented:
@st-jenkins run integration tests please

AudriusButkevicius commented:
@st-jenkins run integration tests please

Zillode commented:
@st-jenkins run integration tests for windows and osx please

AudriusButkevicius commented:
The job itself seems to be broken.

calmh commented on cmd/syncthing/connections.go:
```diff
@@ -266,45 +255,103 @@ func (s *connectionSvc) connect() {
+			}
+
+			last, ok := s.lastRelayCheck[deviceID]
+			if ok && last.Add(time.Duration(s.cfg.Options().RelayReconnectIntervalM)*time.Minute).After(time.Now()) {
```
```go
reconIntv := time.Duration(s.cfg.Options().RelayReconnectIntervalM)*time.Minute
if time.Since(last) > reconIntv {
```

calmh commented on cmd/syncthing/connections.go:
```diff
@@ -266,45 +255,103 @@ func (s *connectionSvc) connect() {
 				}
 
-				conn, err := net.DialTCP("tcp", nil, raddr)
+				to := fmt.Sprintf("%s:%d", net.IP(inv.Address).String(), inv.Port)
```
This would be nice as `SessionInvitation.AddressString()` or so, or just a `SessionInvitation.String()` that makes sense so we could just drop `%v` for the `inv` in the debug prints below..

calmh commented on internal/discover/discover.go:
```diff
@@ -424,30 +485,97 @@ func (d *Discoverer) filterCached(c []CacheEntry) []CacheEntry {
+		}
+	}
+
+	sort.Sort(relayList(relays))
```
This modifies and sort the relay list in place... I'm not sure if that's an issue or perhaps should just be clarified by the comment and not returning anything? The []relay to []string conversion seems orthogonal...

calmh commented on internal/model/connection.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+const (
+	ConnectionTypeBasicAccept ConnectionType = 1 << iota
```
Should this be just ` = iota` unless we use the individual values as bits that can be on and off regardless of each other?

calmh commented:
Do we need an indicator in the GUI for whether a connection is direct or via relay? Help text update on the address field per device to mentioned the "tcp://" prefix (and hostname, while we're at it...)...?

AudriusButkevicius commented:
I am sure I modified the help text etc, though somehow lost it while rebasing.
Yeah, change device address to show the connection type next to it in the UI.

AudriusButkevicius commented:
Performed the following changes:
1 Reworked the incoming relay service. It's now available at the main() scope (rather than connectionSvc scope) see No 3 and 4 as to why.
2. Now shows Relays 0/1 like for global ann.
3. Discovery server now only announces relays which we were able to connect to (perhaps we should follow up with a modification to discosrv to remove all old relays upon receiving an announcement)
4. Moved the relaySvc to a separate package (as now I need to pass it into discovery package) which caused a few other things to be shifted around (such as intermediateConnection and setTcpOptions)

calmh commented:
@st-jenkins please retest this

calmh commented:
@st-jenkins retest this please

calmh commented on internal/relay/relay.go:
```diff
@@ -0,0 +1,197 @@
+	cfg    *config.Wrapper
+	tlsCfg *tls.Config
+
+	receiverToken suture.ServiceToken
```
Unused as far as I can tell?

calmh commented:
WTF? Y u not build...

calmh commented:
Ah, some conflict...

So this looks good to me, possibly there is some opportunity somewhere to reduce imports between model and relay and so on by using an interface or something to cover the IntermediateConnection but I can't pinpoint it right now and it's not essential. I suggest that we get this merged soonish and thus make `master` a pre-v0.12 branch, with maintenance for v0.11 on another branch, as we've done before.

calmh commented:
Something weird with the GUI:

<img width="553" alt="screen shot 2015-07-23 at 10 28 24" src="https://cloud.githubusercontent.com/assets/125426/8846243/a65a71be-3125-11e5-9b94-b5e74b308994.png">

The layout around global discovery server is messed up, but that's already in `master`. However it's now greyed out and not editable?

calmh commented:
I think there's a problem that results in relays not being included in the global discovery announcement. From my reading, the list of relays is taken when the announcement packet is created, which only happens once - when global discovery is started. But global discovery starts before the relays connect, so there are no relays active at that point in time.

```
jb@syno:~/src/github.com/syncthing/syncthing/test (relay) $ STTRACE=net,relay syncthing -home h1
[monitor] 2015/07/23 11:41:43.770100 monitor.go:94: INFO: Starting syncthing
[I6KAH] 2015/07/23 11:41:44.134805 main.go:475: INFO: syncthing v0.11.15+17-g28d00d4-dirty (go1.4.2 darwin-amd64 default) jb@syno 2015-07-22 22:42:08 UTC
[I6KAH] 2015/07/23 11:41:44.134970 main.go:476: INFO: My ID: I6KAH76-66SLLLB-5PFXSOA-UFJCDZC-YAOMLEK-CP2GB32-BV5RQST-3PSROAU
[I6KAH] 2015/07/23 11:41:44.138267 main.go:768: INFO: Database block cache capacity 65536 KiB
[I6KAH] 2015/07/23 11:41:44.145748 model.go:151: INFO: Starting deadlock detector with 20m0s timeout
[I6KAH] 2015/07/23 11:41:44.147448 main.go:636: OK: Ready to synchronize s12 (read-write)
[I6KAH] 2015/07/23 11:41:44.148995 main.go:636: OK: Ready to synchronize default (read-write)
[I6KAH] 2015/07/23 11:41:44.149131 main.go:823: INFO: Starting web GUI on http://127.0.0.1:8081/
[I6KAH] 2015/07/23 11:41:44.149693 rwfolder.go:317: INFO: Completed initial scan (rw) of folder s12
[I6KAH] 2015/07/23 11:41:44.150431 rwfolder.go:317: INFO: Completed initial scan (rw) of folder default
[I6KAH] 2015/07/23 11:41:44.494189 connections.go:92: DEBUG: listening on tcp://127.0.0.1:22001
[I6KAH] 2015/07/23 11:41:44.494250 relay.go:100: DEBUG: Connecting to relay relay://a.relays.syncthing.net:22067
[I6KAH] 2015/07/23 11:41:44.494274 relay.go:100: DEBUG: Connecting to relay relay://b.relays.syncthing.net:22067
[I6KAH] 2015/07/23 11:41:44.494334 main.go:898: INFO: Starting local discovery announcements
[I6KAH] 2015/07/23 11:41:44.494582 connections.go:309: DEBUG: Trying relay connections to JMFJCXB-GZDE4BN-OCJE3VF-65GYZNU-AIVJRET-3J6HMRQ-AUQIGJO-FKNHMQU []
[I6KAH] 2015/07/23 11:41:44.494637 broadcast.go:171: WARNING: Local discovery over IPv4 unavailable: listen udp4 :21025: bind: address already in use
[I6KAH] 2015/07/23 11:41:44.495311 main.go:903: INFO: Starting global discovery announcements
[I6KAH] 2015/07/23 11:41:44.495702 main.go:690: INFO: Device I6KAH76-66SLLLB-5PFXSOA-UFJCDZC-YAOMLEK-CP2GB32-BV5RQST-3PSROAU is "s1" at [dynamic]
[I6KAH] 2015/07/23 11:41:44.495807 main.go:690: INFO: Device JMFJCXB-GZDE4BN-OCJE3VF-65GYZNU-AIVJRET-3J6HMRQ-AUQIGJO-FKNHMQU is "s2" at [dynamic]
[I6KAH] 2015/07/23 11:41:44.495837 main.go:725: INFO: No automatic upgrades; v0.11.15+17-g28d00d4-dirty is not a release version.
[I6KAH] 2015/07/23 11:41:44.505459 gui.go:221: INFO: API listening on 127.0.0.1:8081
[I6KAH] 2015/07/23 11:41:44.550364 client.go:73: DEBUG: ProtocolClient@0xc208ad0480 connected 194.126.249.21:22067
[I6KAH] 2015/07/23 11:41:44.554907 client.go:88: DEBUG: ProtocolClient@0xc208ad0480 joined 194.126.249.21:22067 via 172.21.21.222:57684
[I6KAH] 2015/07/23 11:41:44.633863 client.go:73: DEBUG: ProtocolClient@0xc208ad0360 connected 134.184.43.12:22067
[I6KAH] 2015/07/23 11:41:44.656558 client.go:88: DEBUG: ProtocolClient@0xc208ad0360 joined 134.184.43.12:22067 via 172.21.21.222:57685
```

With the relays not announced, noone tries to connect to this device via the relays...

AudriusButkevicius commented:
That's possible, I'll look into this, or defer global server announcement by n seconds, as it doesn't need to be instant.

calmh commented:
I have a feeling the various announcement senders should take a `pkt []byte` but rather some pointer object that has a method to create an announcement packets, based on the current truth at the time the packet is needed.

AudriusButkevicius commented:
I also want to implement lookup from public pool before we merge, haven't yet seen if you made any comments on IRC as not at home yet.

AudriusButkevicius commented:
Why? That would result in stale data in the byte slice? Or perhaps I am misunderstanding what you mean.

calmh commented:
It's stale now, but should ideally be generated when it's needed. I think we're saying the same thing. :)

I was thinking about that as well. Currently I have added `a.relays.syncthing.net` and `b.relays.syncthing.net` as the two test relays, with `relays.syncthing.net` pointing to them both. It would probably be optimal if we shipped with the code looking up `relays.syncthing.net` and using all the records from there as the list of relays...

AudriusButkevicius commented:
I've explained on IRC why dns does not work, or rather should not be used. Any chance for https://relays.syncthing.net serving a json list of uris?

calmh commented:
I must have missed that explanation - can you repeat it for posterity? But yes, for sure we can server the list over HTTPS instead. Possibly the added security there adds something useful?

---

Edit: I guess one reason is that the Android thingy doesn't currently speak DNS... We have a hardcoded fallback address for data.syncthing.net, so we could do the same for relays.syncthing.net with your scheme.

calmh commented:
This OK?

```
jb@syno:~ $ curl https://relays.syncthing.net/
{
	"relays": [
		{"url": "relay://134.184.43.12:22027"},
		{"url": "relay://194.126.249.21:22027"}
	]
}
```

I'm leaving space for the possibility of other keys than "relays" and "url" (like priority, or bandwidth, or region, or whatever we may have in the future)

AudriusButkevicius commented:
Dns names in urls as such would work. 

I was refering to announcing part via dns srv or multiple a records.

As in, you relays are all A records from relays.syncthing.net

AudriusButkevicius commented:
Now your `[]byte` comment makes sense.

AudriusButkevicius commented:
Closed by accident.

AudriusButkevicius commented:
Changes in the last 2 commits, rebased on master.

AudriusButkevicius commented:
relays.syncthing.net will be contacted on every config save...

calmh commented:
Is there a point to connecting to *all* relays? Would it not suffice to keep a connection to the one with the lowest latency?

calmh commented:
Global announcement changes are neat.

AudriusButkevicius commented:
So my experience is that networks are asymetric. It sometimes might be better to use a more distant relay becuse the other peer has got a better link to it compared to the one close to me.

AudriusButkevicius commented:
Addressed comments.
Regarding connecting to all relays, I am not sure what's right. We can probably change relays.syncthing.net to later round robin balance stuff by returning only one or two of them.

calmh commented:
> It sometimes might be better to use a more distant relay becuse the other peer has got a better link to it compared to the one close to me.

Indeed. I must look closer at how we actually connect via the relays later, I thought we used the one with lowest latency for the announcer -- which should be good enough really, as long as we use one that is best of one of the sides, we should be as close to optimal as we can get with a quick guess

AudriusButkevicius commented:
We can change that for relays from the pool, though explicitly configured ones should stay as they are.

AudriusButkevicius commented:
Whats to be done here?

calmh commented:
Kick back and have a beer while I test it out and find more things to pick on! :dancers:

AudriusButkevicius commented:
:+1 i haven't tested this properly, to be honest. Ran it a while a go for 2 minutes.

calmh commented:
Some things that strike me;

 - Should we have a global discovery as a requirement for doing the relay dance? I don't think the relays are used with global discovery enabled, right?

- There's a minimum time of 10 minutes before we try a relay after startup, as we do the first connect attempt before we have global discovery running, and then remember that we "tried" the relays then. Is this desirable? We do want to prioritize direct connections over relayed, but I think 10 minutes is past most people's "it doesn't work" threshold.

- We announce v4 relays to the v6 only discovery server. I'm not sure this is worth fixing as it's fairly harmless.

calmh commented:
It does connect via relay though \o/

calmh commented on gui/index.html:
```diff
@@ -411,7 +424,11 @@ <h3 class="panel-title">
-                      <th><span class="glyphicon glyphicon-link"></span>&nbsp;<span translate>Address</span></th>
+                      <th>
+                        <span class="glyphicon glyphicon-link"></span>&nbsp;
+                        <span translate ng-if="connections[deviceCfg.deviceID].type.indexOf('basic') == 0" >Address</span>
```
nbsp plus line break here becomes double space

<img width="171" alt="screen shot 2015-07-24 at 16 27 26" src="https://cloud.githubusercontent.com/assets/125426/8876585/e78339a4-3220-11e5-8eb1-2c905108dbd1.png">

uok commented:
Nice!
(btw, what do you think about changing the icon font to [fontawesome](http://fontawesome.io/icons/)?)

AudriusButkevicius commented:
They are advertised via local discovery too. Is there a scenario for a way around corporate fw here? 

I don't understand the 10m point. I though we connect immdeiately? We will always prioritize a direct connection over a relay one, keep trying to dial direct if connected via relay.

I feel that we should awlays announce everything everywhere. V4 and V6 over both, so that there are more chanses to connect, no matter the protocol.

AudriusButkevicius commented:
1. We no longer count `when we last contacted X's relays` until discovery actually returns the first relay for X, giving time GlobalAnn to do it's job.
2. Removed whitespace.
3. Explicit option to start relaySvc when running without GlobalAnn
4. Option not to use relays at all (this is to prevent the connection side too).
5. Rebased on master.

calmh commented:
> They are advertised via local discovery too. Is there a scenario for a way around corporate fw here?

I doubt it... If we can see local announcement we should be able to connect directly as well.

> I don't understand the 10m point. I though we connect immdeiately?

We do. But we do it before global discovery has started, so the connect attempt basically sees "dynamic, no addresses found, no relays found" and fails to connect. But it sets the last attempt time, so we don't retry relays (that we might find in 10 seconds or so when global discovery is up) until relayReconnectTime (10 minutes) later.

AudriusButkevicius commented:
So I've addressed the 10m issue, and I've disabled the relaySvc is no GlobalAnn is used (unless you explicitly flip the switch).

AudriusButkevicius commented:
Let me know if there is srill something left.

calmh commented:
Sorry, I've just been out at a customer doing real work and had no time to look at this. Will get to it soonish. :)

calmh commented:
This works for me. It needs a big painful rebase though. Lets do that and merge it pronto?

calmh commented:
(Then we can work out any remaining kinks in a beta or whatever, but having it in master would be nice)

AudriusButkevicius commented:
Will do it tonight

AudriusButkevicius commented:
Here, hour of my life wasted.
godep is shit.

AudriusButkevicius commented:
@st-jenkins run integration tests please

Zillode commented on lib/config/config.go:
```diff
@@ -467,6 +475,38 @@ func convertV10V11(cfg *Configuration) {
+		}
+	}
+
+	// Use new discovery server
```
@Nutomic incoming change!

Nutomic replied:
So if I read this right, we have to change the global announce server port from 22026 to 22027?

I'll be so happy when the damn DNS works on Android -.-

AudriusButkevicius replied:
The format of listen addresses and device addresses changes too.

Nutomic replied:
Right, but I don't think we use those directly for the Android app, so it shouldn't affect us.

Anyway, I'll keep it in mind to test everything.
